File: backend/pipeline/languages.py
SUPPORTED: dict[str, str] = {
    "eng_Latn": "English",
    "ind_Latn": "Indonesian",
    "zsm_Latn": "Malay",
    "zho_Hans": "Chinese (Simplified)",
    "zho_Hant": "Chinese (Traditional)",
    "jpn_Jpan": "Japanese",
    "kor_Hang": "Korean",
    "tam_Taml": "Tamil",
    "hin_Deva": "Hindi",
    "hin_Latn": "Hindi (romanized)",
}

CONFUSION_GROUPS: dict[str, frozenset[str]] = {
    "malay_indonesian": frozenset({"ind_Latn", "zsm_Latn"}),
    "han": frozenset({"cmn_Hani", "yue_Hani", "zho_Hans", "zho_Hant"}),
    "romanized_hindi": frozenset({"hin_Latn", "urd_Latn", "eng_Latn"}),
}


def group_of(label: str) -> str | None:
    for name, members in CONFUSION_GROUPS.items():
        if label in members:
            return name
    return None


def display_name(label: str) -> str:
    if label == "und":
        return "Undetermined"
    return SUPPORTED.get(label, label)


def script_of_label(label: str) -> str:
    return label.split("_", 1)[1] if "_" in label else ""

File: frontend/src/lib/languages.ts
import type { Stage } from "./types";

// The ten labels backend/pipeline/languages.py supports. Anything else is shown as its raw label.
export const SUPPORTED_LANGUAGES: Record<string, string> = {
  eng_Latn: "English",
  ind_Latn: "Indonesian",
  zsm_Latn: "Malay",
  zho_Hans: "Chinese (Simplified)",
  zho_Hant: "Chinese (Traditional)",
  jpn_Jpan: "Japanese",
  kor_Hang: "Korean",
  tam_Taml: "Tamil",
  hin_Deva: "Hindi",
  hin_Latn: "Hindi (romanized)",
};

export const SCRIPT_NAMES: Record<string, string> = {
  Latn: "Latin",
  Hans: "Han (Simplified)",
  Hant: "Han (Traditional)",
  Hani: "Han",
  Jpan: "Japanese",
  Hang: "Hangul",
  Taml: "Tamil",
  Deva: "Devanagari",
  Arab: "Arabic",
  Cyrl: "Cyrillic",
  Grek: "Greek",
  Thai: "Thai",
};

export const STAGE_LABELS: Record<Stage, string> = {
  general: "General detector",
  specialist: "Specialist detector",
  rule: "Chinese script rule",
};

export function languageName(label: string): string {
  if (label === "und") {
    return "Undetermined";
  }
  return SUPPORTED_LANGUAGES[label] ?? label;
}

export function scriptOf(label: string): string {
  const separator = label.indexOf("_");
  return separator === -1 ? "" : label.slice(separator + 1);
}

export function scriptName(script: string): string {
  return SCRIPT_NAMES[script] ?? script;
}
